subsys/modbus/modbus_serial.c

```diff
@@ -393,12 +393,22 @@ static void cb_handler_tx(struct modbus_context *ctx)
 				   cfg->uart_buf_ctr);
 		cfg->uart_buf_ctr -= n;
 		cfg->uart_buf_ptr += n;
-	} else {
-		/* Disable transmission */
-		cfg->uart_buf_ptr = &cfg->uart_buf[0];
-		modbus_serial_tx_off(ctx);
-		modbus_serial_rx_on(ctx);
+		return;
```

Comment:
Can you please change it to this? What, IMO, in this case looks a little readable.

``` diff c
diff --git a/subsys/modbus/modbus_serial.c b/subsys/modbus/modbus_serial.c
index 2c7cb15e027..4dcc602885c 100644
--- a/subsys/modbus/modbus_serial.c
+++ b/subsys/modbus/modbus_serial.c
@@ -366,7 +366,14 @@ static void cb_handler_tx(struct modbus_context *ctx)
                                   cfg->uart_buf_ctr);
                cfg->uart_buf_ctr -= n;
                cfg->uart_buf_ptr += n;
-       } else {
+               return;
+       }
+
+       /* Must wait till the transmission is complete or
+        * RS-485 tranceiver could be disabled before all data has
+        * been transmitted and message will be corrupted.
+        */
+       if (uart_irq_tx_complete(cfg->dev)) {
                /* Disable transmission */
                cfg->uart_buf_ptr = &cfg->uart_buf[0];
                modbus_serial_tx_off(ctx);
```

Reply:
@mcscholtz It does not look right, please recheck :-)

Reply:
My bad, it should be good now.